File: code/data_structures/stack/stack/stack.ex
# Part of Cosmos by OpenGenus Foundation

# Pattern matching
# new_element = head
# items = tail
defmodule Stack do
  def new, do: [] 
  
  def push(new_element, items), do: [new_element | items]
  
  def pop([element | items]), do: {element, items}
  
  def empty?([]), do: true
  def empty?(_), do: false
  
  def peek([head | _]), do: head
end

File: code/data_structures/src/stack/stack/stack.erl
% Part of Cosmos by OpenGenus Foundation

% Pattern matching
% H -> head
% T -> tail
-module(stack).
-export([new/0, push/2, pop/1, empty/1, peek/1]).

new() -> [].

push(H, T) -> [H, T].

pop([H, T]) -> {H, T}.

empty([]) -> true;
empty(_) -> false.

peek([H, _]) -> H.